File: jApp/src/main/java/j/s/yarlykov/data/db/TabForecast.java
package j.s.yarlykov.data.db;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import j.s.yarlykov.data.domain.CityForecast;

public class TabForecast {

    private final static String TAB_NAME = "forecast";
    private final static String COL_ID = "id";
    private final static String COL_CITY = "city";
    private final static String COL_COUNTRY = "country";
    private final static String COL_TEMP = "temperature";
    private final static String COL_ICON = "icon";
    private final static String COL_WIND = "wind";
    private final static String COL_HUMIDITY = "humidity";
    private final static String COL_PRESSURE = "pressure";

    public static void createTable(SQLiteDatabase db) {

        String query = "CREATE TABLE " + TAB_NAME + " (" +
                COL_ID + " INTEGER PRIMARY KEY NOT NULL, " +
                COL_CITY + " TEXT NOT NULL, " +
                COL_COUNTRY + " TEXT  NOT NULL DEFAULT 'ru', " +
                COL_TEMP + " INTEGER NOT NULL DEFAULT (0), " +
                COL_ICON + " INTEGER NOT NULL DEFAULT (0), " +
                COL_WIND + " REAL NOT NULL DEFAULT (0.0), " +
                COL_HUMIDITY + " INTEGER NOT NULL DEFAULT (0), " +
                COL_PRESSURE + " INTEGER NOT NULL DEFAULT (0))";

        db.execSQL(query);
    }

    public static long addForecast(CityForecast cf, SQLiteDatabase db) {
        return db.insert(TAB_NAME, null, initContentValues(cf));
    }

    public static int editForecast(CityForecast cf, SQLiteDatabase db) {
        return db.update(TAB_NAME, initContentValues(cf),
                COL_CITY + "='" + cf.getCity() + "'", null);
    }

    public static CityForecast getCityForecast(String city, SQLiteDatabase db) {

        try(Cursor cursor = db.rawQuery("SELECT * FROM " + TAB_NAME + " WHERE " +
                COL_CITY + "=? LIMIT 1", new String[]{city})) {

            if (cursor.moveToFirst()) {
                return extractForecast(cursor);
            }
        }
        return null;
    }

    private static ContentValues initContentValues(CityForecast cf) {
        ContentValues cv = new ContentValues();
        cv.put(COL_CITY, cf.getCity());
        cv.put(COL_COUNTRY, cf.getCountry());
        cv.put(COL_ICON, cf.getIcon());
        cv.put(COL_TEMP, cf.getTemperature());
        cv.put(COL_WIND, cf.getWind());
        cv.put(COL_HUMIDITY, cf.getHumidity());
        cv.put(COL_PRESSURE, cf.getPressure());
        return cv;
    }

    private static CityForecast extractForecast(Cursor cursor) {
        int idxId = cursor.getColumnIndex(COL_ID);
        int idxCity = cursor.getColumnIndex(COL_CITY);
        int idxCountry = cursor.getColumnIndex(COL_COUNTRY);
        int idxTemp = cursor.getColumnIndex(COL_TEMP);
        int idxIcon = cursor.getColumnIndex(COL_ICON);
        int idxWind = cursor.getColumnIndex(COL_WIND);
        int idxHumidity = cursor.getColumnIndex(COL_HUMIDITY);
        int idxPressure = cursor.getColumnIndex(COL_PRESSURE);

        return new CityForecast(
                cursor.getString(idxCity),
                cursor.getString(idxCountry),
                cursor.getInt(idxTemp),
                cursor.getInt(idxIcon),
                cursor.getFloat(idxWind),
                cursor.getInt(idxHumidity),
                cursor.getInt(idxPressure));
    }
}

File: kApp/src/main/java/k/s/yarlykov/data/db/TabForecast.kt
package k.s.yarlykov.data.db

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import k.s.yarlykov.data.domain.CityForecast

object TabForecast {

    val TAB_NAME = "forecast"
    val COL_ID = "id"
    val COL_CITY = "city"
    val COL_COUNTRY = "country"
    val COL_TEMP = "temperature"
    val COL_ICON = "icon"
    val COL_WIND = "wind"
    val COL_HUMIDITY = "humidity"
    val COL_PRESSURE = "pressure"

    fun createTable(db: SQLiteDatabase?) {

        val query = "CREATE TABLE $TAB_NAME (" +
                "$COL_ID INTEGER PRIMARY KEY NOT NULL, " +
                "$COL_CITY STRING NOT NULL, " +
                "$COL_COUNTRY STRING  NOT NULL DEFAULT 'ru', " +
                "$COL_TEMP INTEGER NOT NULL DEFAULT (0), " +
                "$COL_ICON INTEGER NOT NULL DEFAULT (0), " +
                "$COL_WIND REAL NOT NULL DEFAULT (0.0), " +
                "$COL_HUMIDITY INTEGER NOT NULL DEFAULT (0), " +
                "$COL_PRESSURE INTEGER NOT NULL DEFAULT (0))"

        db?.execSQL(query)
    }

    // Добавить запись
    fun addForecast(f: CityForecast, db: SQLiteDatabase) : Long {
        return db.insert(TAB_NAME, null, initContentValues(f))
    }

    // Изменить запись
    fun editForecast(f: CityForecast, db: SQLiteDatabase): Int {
        return db.update(TAB_NAME, initContentValues(f), "$COL_CITY='${f.city}'", null)
    }

    fun deleteById(id: Int, db: SQLiteDatabase) {
        db.delete(TAB_NAME, "$COL_ID=$id", null)
    }

    fun deleteByCity(city: String, db: SQLiteDatabase) {
        db.delete(TAB_NAME, "$COL_CITY=$city", null)
    }

    fun deleteAll(db: SQLiteDatabase) {
        db.delete(TAB_NAME, null, null)
    }

    fun getCityForecast(city: String, db: SQLiteDatabase) : CityForecast? {

        var forecast: CityForecast? = null
        val cursor: Cursor = db.rawQuery("SELECT * FROM $TAB_NAME WHERE $COL_CITY=? limit 1", arrayOf(city))

        if(cursor.moveToFirst()) {
            forecast = extractForecast(cursor)
        }

        cursor.close()
        return forecast
    }

    fun readAll(db: SQLiteDatabase): List<CityForecast> {
        val list = mutableListOf<CityForecast>()

        val cursor = db.query(TAB_NAME, null, null, null, null, null, null);

        if(cursor.moveToFirst()) {
            do{
                list.add(extractForecast(cursor))
            } while (cursor.moveToNext())
        }
        cursor.close()
        return list
    }

    private fun extractForecast(cursor: Cursor) : CityForecast {

        val idxId = cursor.getColumnIndex(COL_ID)
        val idxCity = cursor.getColumnIndex(COL_CITY)
        val idxCountry = cursor.getColumnIndex(COL_COUNTRY)
        val idxTemp = cursor.getColumnIndex(COL_TEMP)
        val idxIcon = cursor.getColumnIndex(COL_ICON)
        val idxWind = cursor.getColumnIndex(COL_WIND)
        val idxHumidity = cursor.getColumnIndex(COL_HUMIDITY)
        val idxPressure = cursor.getColumnIndex(COL_PRESSURE)

        return CityForecast(
                cursor.getInt(idxId),
                cursor.getString(idxCity),
                cursor.getString(idxCountry),
                cursor.getInt(idxTemp),
                cursor.getInt(idxIcon),
                cursor.getFloat(idxWind),
                cursor.getInt(idxHumidity),
                cursor.getInt(idxPressure))
    }

    private fun initContentValues(f: CityForecast): ContentValues {
        return ContentValues().apply {
            put(COL_CITY, f.city)
            put(COL_COUNTRY, f.country)
            put(COL_ICON, f.icon)
            put(COL_TEMP, f.temperature)
            put(COL_WIND, f.wind)
            put(COL_HUMIDITY, f.humidity)
            put(COL_PRESSURE, f.pressure)
        }
    }
}